### PR TITLE
lorawan: add downlink reading capability

### PR DESCRIPTION
**Pull request by KubaFYI**
This works either by holding the latest N received messages
in a buffer. Normally N=1 but optionally a larger buffer
(using ring buffer) can be enabled.

This commit also removes a compile warning from lorawan_init and
adds a board overlay for Adafruit Feather M0 Lora version to the
LoRaWAN example code.

Signed-off-by: Kuba Sanak <contact@kuba.fyi>

### CODE REVIEW

**Comment by KubaFYI** [comment]
I'm not married to the exact form of this API, so quite happy to take input on it.

In particular I am aware that it might be unnecessary to have the buffering of downlink messages to be handled by the lorawan subsystem itself and it could be done by user app, but I wanted to avoid a situation where a downlink would be missed because it's not been collected by the user before another one arrives (e.g. in class C operation). Though now than I think about it perhaps a neater way to deal with that would be to allow for registration of a callback upon arrival of a downlink message?

**Comment by adityaxavier** [comment]
@KubaFYI In my opinion it might be better to have an API to register a Callback function for Receiving and receive the payload in the receiving function. Similar to registering for BLE / Events..

**Comment by Mani-Sadhasivam** [comment]
> Though now than I think about it perhaps a neater way to deal with that would be to allow for registration of a callback upon arrival of a downlink message?

@KubaFYI Yes, I really like to have the callback way of receiving the data. Then we can offload the data handling to the user application and that will be very convenient.

**Comment by KubaFYI** [comment]
Cool, that makes sense. I'll include a callback interface. Should I remove the non-callback parts then?

**Comment by Mani-Sadhasivam** [comment]
> Should I remove the non-callback parts then?

Yes, please. It'd be good if you can show some basic way of handling the data in sample application like having a ring buffer which you're using in loramac currently.

**Comment by KubaFYI** [comment]
That's done now.

File: scripts/west_commands/runners/bossac.py [hunk 1]
```diff
@@ -98,6 +98,7 @@ def get_offset(self, supports_offset):
                 ' Please see' +
                 ' https://github.com/zephyrproject-rtos/sdk-ng/issues/234' +
                 ' which tracks updating the Zephyr SDK.')
+            return None
```
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/bossac.py:101]: This belongs in a separate commit.

**Review by mbolivar-nordic — DISMISSED** [review]
Please split the west bossac runner change into its own commit. CC @nandojve @nzmichaelh

File: scripts/west_commands/runners/bossac.py (hunk 1 above)
Comment (KubaFYI) [reply scripts/west_commands/runners/bossac.py]: Ah apologies, unintentional leftover from my private test setup.

File: samples/lorawan/class_a/src/main.c [hunk 2]
```diff
@@ -25,13 +25,32 @@ BUILD_ASSERT(DT_NODE_HAS_STATUS(DEFAULT_RADIO_NODE, okay),
...
+	static char hex_str[DL_MSG_LEN*2+1];
+	int i;
+
+	for (i = 0; i < length; i++) {
+		hex_str[i*2] = (payload[i] & 0xF0) >> 4;
+		hex_str[i*2] = hex_str[i*2] < 10 ?
+			'0' + hex_str[i*2] : 'A' + hex_str[i*2];
+		hex_str[i*2+1] = payload[i] & 0x0F;
+		hex_str[i*2+1] = hex_str[i*2+1] < 10 ?
+			'0' + hex_str[i*2+1] : 'A' + hex_str[i*2+1] - 10;
```
Comment (Mani-Sadhasivam) [inline samples/lorawan/class_a/src/main.c:47]: Can't you use `bin2hex`?

**Review by Mani-Sadhasivam — CHANGES_REQUESTED** [review]

File: include/lorawan/lorawan.h [hunk 3]
```diff
@@ -99,6 +99,20 @@ struct lorawan_join_config {
 	enum lorawan_act_type mode;
 };
 
+/**
+ * @brief Register a callbact to be called in the case of a downlink message
+ *
+ * The callback function takes three parameters:
+ *	- port - port number of incoming message (uint8_t)
+ *	- length - length of the incoming in bytes (uint8_t)
+ *	- payload - uint8_t array containint the incoming message
```
[the remark this answers is not in the document]
Comment (KubaFYI) [reply include/lorawan/lorawan.h]: Fixed this and a number of other misspellings as well.

File: samples/lorawan/class_a/src/main.c (hunk 2 above)
Comment (KubaFYI) [reply samples/lorawan/class_a/src/main.c]: Didn't realise that util existed. Using it now.

File: samples/lorawan/class_a/src/main.c [hunk 4]
```diff
@@ -25,13 +26,27 @@ BUILD_ASSERT(DT_NODE_HAS_STATUS(DEFAULT_RADIO_NODE, okay),
...
 #include <logging/log.h>
 LOG_MODULE_REGISTER(lorawan_class_a);
 
 char data[] = {'h', 'e', 'l', 'l', 'o', 'w', 'o', 'r', 'l', 'd'};
 
+void downlink_printer(uint8_t port, uint8_t length, uint8_t *payload)
+{
+	char hex_str[DL_MSG_LEN*2+1];
+	int ret;
+	ret = bin2hex(hex_str, DL_MSG_LEN*2+1, payload, length);
```
Comment (aurel32) [inline samples/lorawan/class_a/src/main.c:41]: You might want to use `LOG_HEXDUMP_INF` which does all the job of converting the data to hex format.

**Review by Mani-Sadhasivam — APPROVED** [review]
LGTM! Thanks.

**Comment by andysan** [comment]
Stupid question... :) Is this really the API we want? Would it be more convenient to register callbacks per port? That way, we could neatly separate different well-known protocols (e.g., the test interface).

A long time ago, I hacked up this that does just that: https://github.com/andysan/zephyr/commit/8b5a518ec7cd42aab95292489f27fa4223b1592b

**Comment by KubaFYI** [comment]
> Is this really the API we want? Would it be more convenient to register callbacks per port? That way, we could neatly separate different well-known protocols (e.g., the test interface).

Is there a reason why this shouldn't be done by the application if it has a need for it? Let's say I needed the same callback to be executed for a hundred different ports - as a user it'd be a bit cumbersome to have to register a callback for each one of those separately instead of just registering one callback with a switch statement on a port number.

File: samples/lorawan/class_a/src/main.c (hunk 4 above)
Comment (KubaFYI) [reply samples/lorawan/class_a/src/main.c:41]: I wanted to be able to format the log string with the port number the message arrived as and `LOG_HEXDUMP_INF` doesn't offer string formatting afaic?

**Comment by andysan** [comment]
> > Is this really the API we want? Would it be more convenient to register callbacks per port? That way, we could neatly separate different well-known protocols (e.g., the test interface).
> 
> Is there a reason why this shouldn't be done by the application if it has a need for it? Let's say I needed the same callback to be executed for a hundred different ports - as a user it'd be a bit cumbersome to have to register a callback for each one of those separately instead of just registering one callback with a switch statement on a port number.

The main reason is that it makes it easier to write modular code. There are several standardised ports used for things like testing and multicast that should be handled outside most applications. It's the exact same reason you need to bind specific ports in the IP stack if you want to accept incoming packets.

My implementation has a mechanism to handle the case you mentioned, you just register a listener by calling `lorawan_listen(LW_RECV_PORT_ANY, &my_callback)`. That listener then gets called for any port that doesn't have a separate handler registered.

**Comment by Mani-Sadhasivam** [comment]
> My implementation has a mechanism to handle the case you mentioned, you just register a listener by calling `lorawan_listen(LW_RECV_PORT_ANY, &my_callback)`. That listener then gets called for any port that doesn't have a separate handler registered.

@KubaFYI I think this makes sense! I didn't think about the need for this and @andysan got it right. Please incorporate the suggestion.

**Comment by JordanYates** [comment]
Are there any objections to extending this callback so that it is run on every call to `McpsIndication`?
The primary reason being that even downlink packets without data still contain useful information (The network server is still able to communicate with us). It also more accurate to the name of the API, `downlink_callback` not `downlink_with_data_callback`.
Empty packets already have `Port=0`, `Buffer=NULL`, and `BufferSize=0`, so no special casing is required.
